dbschema/project.gel

```diff
@@ -208,4 +303,65 @@ module Project {
       on target delete allow;
     };
   }
+
+  # creates the ranges for the given start and end dates based upon the given month interval
+  function create_periodic_report_ranges(startDate: cal::local_date, endDate: cal::local_date, 
+    monthInterval: str) -> set of range<cal::local_date>
+    using (
+      with
+        reportingPeriod := range(<cal::local_date>startDate, <cal::local_date>endDate),
+        reportPeriodStartDates := range_unpack(reportingPeriod, <cal::date_duration>(monthInterval ++ ' month')),
+        reportPeriodRanges := (
+          for firstDayOfMonth in reportPeriodStartDates
+          union (
+            with
+              firstDayOfNextMonth := (select firstDayOfMonth + <cal::relative_duration>(monthInterval ++ ' month')),
+              lastDayOfMonth := firstDayOfNextMonth - <cal::relative_duration>'1 day'
+            select range(<cal::local_date>firstDayOfMonth, <cal::local_date>lastDayOfMonth)
+          ))
+      select reportPeriodRanges
+  );
+
+  function insertReports(requestedReportPeriods: array<range<cal::local_date>>,
+    existingReports: array<default::PeriodicReport>, 
+    newProject: default::Project) -> set of default::PeriodicReport
+    using (
+      with
+        requestedReportPeriodsForInsertion := (
+          select array_unpack(requestedReportPeriods)
+          filter array_unpack(requestedReportPeriods) not in (
+            for report in array_unpack(existingReports)
+            union (
+              select report.period
+            )
+          )
+        ),
```

Comment:
@CarsonF - I think this whole thing is pretty close, but on this statement here I'm receiving this warning:

`possibly more than one element returned by an expression in a FILTER clause`

on line 335.  Any ideas on how to remedy that?

Reply:
I tried using an `assert_single`, but I'm not even sure if that's safe.

Reply:
This seems to take care of the the warning:

```edgeql
        distinctRequestedReportPeriods := distinct array_unpack(requestedReportPeriods),
        requestedReportPeriodsForInsertion := (
          select distinctRequestedReportPeriods
          filter distinctRequestedReportPeriods not in (
            for report in array_unpack(existingReports)
            select report.period
          )
        ),
```